Batch executemany

fantix opened the pull request:
Refs #289, this is a rewrite of `bind_execute_many()`:

* Batch bind args into 4 x 32KB buffers, sent with `writelines()` once flow control turns green.
* Send one `SYNC` at the very last - making use of the implicit transaction for the whole call to `executemany()`.
* Abort immediately on either client error (with explicit rollback) or server error.
* ~Support async iterable.~
* Add `executemany()` to prepared statement.

- [x] `test_server_failure_during_writes` is failing on Windows
- [x] Fix unstable `test_timeout (test_execute.TestExecuteMany)`

**This is a breaking change due to the implicit transaction.**

![executemany](https://user-images.githubusercontent.com/1751601/49010730-180ee400-f1af-11e8-9759-0a37f5f0ea15.png)

The 3-level infinite loop keeps filling buffers and send them under flow control, the only exit is raising `StopIteration` with 3 possible values:
* `True` means there was data sent in the last loop
* `False` means the last loop sent nothing
* `exception` means the data source raised an exception

Please note, during the loop, if server-side error is detected before exhausting the data source, it is also considered as `StopIteration(True)`, ending the process early with a `SYNC_MESSAGE`.

(I [wrote a post](https://segmentfault.com/a/1190000017136059) about the details of implicit transaction.)

***UPDATED*** - [pgbench](https://github.com/fantix/pgbench/commit/23ef4366b177e5b7c81ef199140db2ddd3f25997) results of inserting 1000 rows per query with `executemany()` on Python 3.6 of 2.2GHz 2015 MacBook Air (best out of 5 runs):

asyncpg 0.18.2

```
710 queries in 30.31 seconds
Latency: min 341.88ms; max 636.29ms; mean 425.022ms; std: 39.782ms (9.36%)
Latency distribution: 25% under 401.67ms; 50% under 414.26ms; 75% under 435.37ms; 90% under 478.39ms; 99% under 576.638ms; 99.99% under 636.299ms
Queries/sec: 23.42
Rows/sec: 23424.32
```

This PR:

```
4125 queries in 30.02 seconds
Latency: min 23.14ms; max 734.91ms; mean 72.723ms; std: 49.226ms (67.69%)
Latency distribution: 25% under 59.958ms; 50% under 65.414ms; 75% under 71.538ms; 90% under 80.95ms; 99% under 175.375ms; 99.99% under 734.912ms
Queries/sec: 137.39
Rows/sec: 137389.64
```

Most branches are covered by tests, overall coverage 85% (1352 miss) -> 85% (1356 miss).

***UPDATED*** - Python 3.8 of 2.3GHz 2020 MacBook Pro:

<img width="938" alt="截屏2020-11-24 17 21 26" src="https://user-images.githubusercontent.com/1751601/100162742-ee5e2580-2e79-11eb-8aae-ef64b946c99b.png">

fantix commented on asyncpg/protocol/coreproto.pyx:
```diff
@@ -796,30 +798,55 @@ cdef class CoreProtocol:
+            self._set_state(PROTOCOL_BIND_EXECUTE_MANY_ROLLBACK)
+        packet = WriteBuffer.new()
+
+        # We have no idea if we are in an explicit transaction or not,
```
Yeah just saw that coming from ReadyForQuery.

fantix commented on asyncpg/protocol/coreproto.pyx:
```diff
@@ -796,30 +798,55 @@ cdef class CoreProtocol:
+        # therefore we raise an exception in the database to mark current
+        # transaction as failed anyway to prevent partial commit
+        buf = self._build_parse_message('',
+                                        'DO language plpgsql $$ BEGIN '
```
Because if we are in a subtransaction (savepoint), ROLLBACK would abort through all subtransactions. Okay in this case maybe it is the only way to e.g. hack in SELECT 1/0? Please kindly advise :)

elprans replied:
I think we should not touch the existing transaction state.  If we are currently in transaction, just do nothing.  An unhandled exception in `executemany` iterator would crash the whole thing, and if they _do_ handle the exception, then the intention is clear and explicit.  We only want to guard against unhandled exceptions.

elprans commented on asyncpg/protocol/coreproto.pyx:
```diff
@@ -810,34 +810,29 @@ cdef class CoreProtocol:
+        # We shall rollback in an implicit transaction to prevent partial
+        # commit, while do nothing in an explicit transaction and leaving the
+        # error to the user
+        if self.xact_status == PQTRANS_IDLE:
```
I'd move `is_in_transaction()` to `coreproto.pyx` (where it belongs) and use that instead here.

1st1 commented on asyncpg/protocol/protocol.pyx:
```diff
@@ -217,29 +220,128 @@ cdef class BaseProtocol(CoreProtocol):
+                                    more = False
+                                    break
+                            else:
+                                args_item = iterator.__next__()
```
Just use the `next()` built-in.

1st1 commented on asyncpg/protocol/protocol.pyx:
```diff
@@ -217,29 +220,128 @@ cdef class BaseProtocol(CoreProtocol):
+        try:
+            aiter = args.__aiter__
+        except AttributeError:
+            iterator = args.__iter__()
```
Just use the `iter()` built-in.

fantix commented on asyncpg/protocol/protocol.pyx:
```diff
@@ -217,29 +220,128 @@ cdef class BaseProtocol(CoreProtocol):
+                        while more and packet.len() < _EXECUTE_MANY_BUF_SIZE:
+
+                            # Grab the next item from input
+                            if async_iterator:
```
There was no particular reason or use case to add this - I saw `copy_in()` have it and it was kinda easy to fit in. Yeah it's a bit tricky when an asynchronous iterable/iterator is at the same time an asynchronous generator - it depends on the scenario requirement whether to `athrow()` an exception on error or not. So if there's no further ideas, I'll remove this first.

elprans replied:
The difference in `copy_in()` is that it is designed to handle very large amounts of data, and the source async iterable is allowed to return an arbitrary amount of data on every iteration.

fantix replied:
Cool, thanks for the explanation and time reviewing! I'll fix the issues we addressed here in another commit. By the way, do you prefer squashing the commits of a PR into a big one, or keeping the details as they are?

elprans replied:
Squashing is preferable. Thanks!

1st1 replied:
> The difference in copy_in() is that it is designed to handle very large amounts of data, and the source async iterable is allowed to return an arbitrary amount of data on every iteration.

Maybe it's a good idea to add this explanation as a comment in `copy_in` implementation?

fantix replied:
Added 😃

fantix commented:
Update - I've met a racing condition issue during fixing tests. Trying to find time to fix that.

elprans commented:
Hey @fantix! Are you still working on this?

fantix commented:
Hi! Yes I’m trying to find some time to fix the racing issue, sorry for the delay!

fantix commented:
Working on this now - rebased and I'll review the changes, and deal with the racing issue.

fantix commented:
Sorry for the delay! This PR is updated now. Benchmark shows the improvement is consistent. I've also added a diagram to explain the details.

elprans commented:
@fantix The `test_timeout` test is flaky because of pipelining.  All commands, including the final `Sync` can actually go in under 500ms, so the changes are persisted despite the `TimeoutError`.  You might get better luck by passing a generator with `time.sleep()` in it to `executemany()`.

elprans commented:
Hey @fantix, I made a fix to the timeout issue here: https://github.com/MagicStack/asyncpg/commit/bbf4d55958b1a85ce48abb4283be9a401f57d4fd  Take a look when you have time, squash into your PR, and and let's merge this!  I'm planning to cut 0.19.0 this week ideally.

fantix commented:
Got it, thanks a lot!

Andrei-Pozolotin commented:
curious: what is missing to make this accepted?

victoraugustolls commented:
@elprans is this okay to be merged? I would be happy to help if necessary, looking forward to it as I have a similar problem as #420 !

gjeusel commented:
In awaiting for the PR to be merged, here is a neat trick to get some performances in bulk upsert: [credits to Schinckel](https://schinckel.net/2019/12/13/asyncpg-and-upserting-bulk-data/)

It consists in:
- creating a temporary table
- copy_records_to_table (binary copy) onto this table
- insert into the original table with the "ON CONFLICT" statement, selecting from the temporary one

Syniex commented:
Is there a reason it doesn't get merged?

robd003 commented:
@elprans any chance this can get merged in after the conflict in asyncpg/protocol/protocol.pyx is fixed?

gtlanigan315 commented:
Checking in to see if this PR is still planning on being merged

elprans commented:
@fantix, are you interested in getting this in shape to be merged?

codyflynnNC commented:
bumping this PR. it would be awesome to see it finally merged!! lets goooo ^^

fantix commented:
Let me try this weekend!

fantix commented:
Only rebased - I do want to revisit this to make sure everything is still alright.

fantix commented:
I've rewritten the core part to be more readable, and included @elprans 's commit to fix the timeout, and added one more test. If the Travis CI passes, I think it'll be good for another review.

robd003 commented:
Looks like all tests passed! Can we get this merged in?

elprans commented:
LGTM and merged.  @fantix, thank you very much for working on this!
